### ISSUE TITLE
What to do about nil/null inputs

### ISSUE THREAD

**Issue by phlogistonjohn** [ceph/go-ceph#427]
This is a topic that frequently comes up in reviews and I don't think we on the maintenance team have a solid consensus around this. I'm creating this issue to kick off some discussion related to the above.

In some cases the callers of functions in the go-ceph API can pass a nil value. This can be when the argument is a either a pointer type or an interface type. In response, the code can do one of three things [1]:
* Panic (Go style)
* Return an error
* Abort/crash (C/C++ style)

In most Go code I generally expect the first, to panic, with a nil de-reference. However, as we wrap a C/C++ library we can produce unexpected (from a Go library perspective) behaviors of crashing. In addition, IIRC in some reviews I think that reviewers have requested the "return an error" behavior, as it is "nicer" from a workflow of a user consuming the library.

I'd like to develop a consistent behavior for the library that can be used as a general approach. This general approach would be tested whenever appropriate.

[1] - If I missed something, please say so.

**Comment by phlogistonjohn** [comment]
My proposal:
Consistently produce panics, never crash.

I prefer this approach because it strikes me as the most similar to what a pure-Go library would do. It would not require "armor-plating" all the calls that accept nil values to produce errors. As as side effect it would not require calls that naturally do not return errors to do so. Every where a nil could be passed to the ceph apis it would be tested to ensure that no nils are passed to C calls inappropriately to cause a crash.

**Comment by phlogistonjohn** [comment]
@nixpanic @ansiwen  I think you both have advocated the "return errors" approach in the past. Let's discuss. :-)

**Comment by phlogistonjohn** [comment]
I should also add that this only covers nil inputs, not "malformed" or "incomplete" values, like an rbd image that was not constructed properly. I consider that a related but separate topic, but if you want to take on some discussion of that here, we can do so too.

**Comment by nixpanic** [comment]
I prefer to not use `null` as an argument to C functions. We have seen undefined behaviour when passing `null` to the C library APIs.

As long as that is prevented, I do not really care much if we panic in the go-ceph function, or if an error is returned.

**Comment by ansiwen** [comment]
I would use the guideline that we can always crash/panic whenever the equivalent C function would/might also crash or has undefined behavior. In doubt I would go for the safer option and the option of least surprise, especially if the check is almost for free.

Regarding your proposal @phlogistonjohn, I don't mind too much if we also crash (you mean segfault etc. right?) or we prevent the crash and return a panic instead, because I would not like to propagate a programming style that uses panics as a kind of exception. So I think we should only panic when there is a programming error or a fatal resource error (like OOM). Let's avoid `recover()` as a part of the API.

So, to sum it up: I'm fine with (but don't demand) panic instead of crash, I'm not fine with panic instead of (non-fatal) error values, that is when you used the API correctly and there is no fatal situation. So if in doubt, upgrade in the direction crash->panic->error.

(Nit: also maps and slices can be nil)

**Comment by phlogistonjohn** [comment]
I think we have a rough consensus (see above). There's no need to keep this open any longer.
